### PR TITLE
Fix safety profile MultiDPTest and BuiltInTopicTest

### PR DESCRIPTION
**Pull request by jonesc-oci**
Safety profile MultiDPTests can fail because RTPS discovery isn't given time to complete. Adding a delay in the subscriber addresses the issue.

### CODE REVIEW

File: tests/DCPS/MultiDPTest/subscriber.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/subscriber.cpp:132]: @mitza-oci I'm still not sure why the RTPS discovery has a delay in safety profile and why the delay is 1 second. I'm curious what's the reason?

Comment (mitza-oci) [reply tests/DCPS/MultiDPTest/subscriber.cpp:132]: There is no built-in delay of 1 second, however participant discovery can take some time depending on how it's configured.  Ideally these tests would have no assumptions about exactly how quickly asynchronous events like discovery take to complete.

I think last time we discussed this the plan was to make another change before merging, is that correct?

Comment (sonndinh) [reply tests/DCPS/MultiDPTest/subscriber.cpp:132]: If I heard it right, Chip is checking if that timing assumption is an issue and worth a more extensive fixing.

Other than that, the CI is having a bunch of errors, but I think that's because this branch isn't up-to-date with the master.

File: tests/DCPS/MultiDPTest/publisher.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/publisher.cpp:226]:
```suggestion
        ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: get_publication_matched_status failed!\n")));
```

File: tests/DCPS/MultiDPTest/subscriber.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/subscriber.cpp:22]:
```suggestion
#include <dds/DCPS/BuiltInTopicUtils.h>
```

File: tests/DCPS/MultiDPTest/subscriber.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/subscriber.cpp:239]: Looks like this variable is unused for builds that are not safety profile.

File: tests/DCPS/MultiDPTest/subscriber.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/subscriber.cpp:267]:
```suggestion
          ACE_ERROR ((LM_ERROR, "(%P|%t) ERROR:  failed to read BIT participant data.\n"));
```

File: tests/DCPS/MultiDPTest/subscriber.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/subscriber.cpp:269]:
```suggestion
          num_participants = part_data.length();
```

File: tests/DCPS/BuiltInTopicTest/monitor.cpp
Comment (sonndinh) [inline tests/DCPS/BuiltInTopicTest/monitor.cpp:176]:
```suggestion
      ACE_OS::sleep(3);
```

File: tests/DCPS/BuiltInTopicTest/monitor.cpp
Comment (sonndinh) [inline tests/DCPS/BuiltInTopicTest/monitor.cpp:295]: Should these lines be removed?

File: tests/DCPS/BuiltInTopicTest/monitor.cpp
Comment (sonndinh) [inline tests/DCPS/BuiltInTopicTest/monitor.cpp:381]: Remove these?

File: tests/DCPS/MultiDPTest/publisher.cpp
Comment (jonesc-oci) [reply tests/DCPS/MultiDPTest/publisher.cpp]: Thanks. I removed all but the publisher finished file which the subscriber so it can read the number of timeout_writes

**Comment by ClaytonCalabrese** [comment]
In tests/dcps_tests.lst we have blockers set up for these two tests in CI.

`tests/DCPS/BuiltInTopicTest/run_test.pl: !DCPS_MIN !NO_BUILT_IN_TOPICS !DDS_NO_OWNERSHIP_PROFILE !TARGET !GH_ACTIONS_OPENDDS_SAFETY_PROFILE`

`tests/DCPS/MultiDPTest/run_test.pl: !DCPS_MIN !DDS_NO_OWNERSHIP_PROFILE !GH_ACTIONS_OPENDDS_SAFETY_PROFILE`|

These are blocked by the `!GH_ACTIONS_OPENDDS_SAFETY_PROFILE` part. You should remove that part in order to run these tests in the CI. Additionally docs/internal/github_actions.rst describes which tests are currently being blocked for certain CI configurations, including these two tests, and should be updated alongside the unblocking of these test.  These two sections can probably be removed:

`tests/DCPS/BuiltInTopicTest/run_test.pl
test_u18_esafe_js0`

`tests/DCPS/MultiDPTest/run_test.pl
test_u18_bsafe_js0_FM-1f
test_u18_esafe_js0`

File: tests/DCPS/MultiDPTest/publisher.cpp
Comment (sonndinh) [inline tests/DCPS/MultiDPTest/publisher.cpp:242]: We should keep this line so that the file is closed before the publisher ends.

Comment (sonndinh) [reply tests/DCPS/MultiDPTest/publisher.cpp:242]: @mitza-oci In the case of RELIABLE_RELIABILITY_QOS, if the publisher exits before all of its samples have been received by the subscriber, will the rest of the samples still be received by the subscriber eventually?

**Comment by sonndinh** [comment]
> 

@jonesc-oci Please update the `tests/dcps_tests.lst` file as commented by Clayton. That will enable these two tests in GitHub Actions; right now they are not run yet.

File: tests/DCPS/BuiltInTopicTest/subscriber.cpp
Comment (iguessthislldo) [inline tests/DCPS/BuiltInTopicTest/subscriber.cpp:189]:
```suggestion
      dr_qos.representation.value[0] = DDS::XCDR2_DATA_REPRESENTATION;
#else
      dr_qos.representation.value[0] = OpenDDS::DCPS::UNALIGNED_CDR_DATA_REPRESENTATION;
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: tests/DCPS/MultiDPTest/publisher.cpp
Comment (jonesc-oci) [reply tests/DCPS/MultiDPTest/publisher.cpp:242]: Yes. Thanks for catching that.

Comment (mitza-oci) [reply tests/DCPS/MultiDPTest/publisher.cpp:242]: > @mitza-oci In the case of RELIABLE_RELIABILITY_QOS, if the publisher exits before all of its samples have been received by the subscriber, will the rest of the samples still be received by the subscriber eventually?

Samples that are already in the reader's cache are still available, but the reader's QoS settings will influence this too.  Samples that have not been transferred to the reader's cache do not outlive the writer unless you use a higher level of Durability (which would not be supported in a Safety Profile build).

Comment (sonndinh) [reply tests/DCPS/MultiDPTest/publisher.cpp:242]: Thanks. So I think there is an issue with these changes. The publisher can exit before the subscriber receives all samples, and if that happens it'll cause the subscriber to timeout. I think we should wait for the subscriber to finish here, for example, using `WaitSet` since we are getting rid of `sub_finished_filename` file that was used to indicate the subscriber's completion. That's the last major comment I have.

**Review by sonndinh — APPROVED** [review]

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**
